parser-typechecker/src/U/Codebase/Branch/Diff.hs

```diff
@@ -172,3 +189,27 @@ nameChanges namePrefix (TreeDiff (DefinitionDiffs {termDiffs, typeDiffs} :< chil
       xs
         & Set.toList
         & fmap (name,)
+
+-- | Get a 'NameBasedDiff' from a 'TreeDiff'.
+nameBasedDiff :: TreeDiff -> NameBasedDiff
+nameBasedDiff (TreeDiff (DefinitionDiffs {termDiffs, typeDiffs} :< children)) =
+  let NameBasedDiff childrenTerms childrenTypes =
+        foldMap (nameBasedDiff . TreeDiff) children
+   in NameBasedDiff
+        { terms = foldMap nameBasedTermDiff termDiffs <> childrenTerms,
+          types = foldMap nameBasedTypeDiff typeDiffs <> childrenTypes
+        }
+  where
+    nameBasedTermDiff :: Diff Referent -> Relation Reference Reference
+    nameBasedTermDiff Diff {adds, removals} =
+      let termAdds = mapMaybe Referent.toTermReference (Set.toList adds)
+          termRemovals = mapMaybe Referent.toTermReference (Set.toList removals)
```

Comment:
Is there a particular reason we don't patch constructors?

I don't think that change would happen in this PR, but it seems like it'd be easy to add here.

Reply:
They just aren't representable in a patch right now, so I left them out of this structure, too.

Old ticket: https://github.com/unisonweb/unison/issues/3504